File: 1056-capacity-to-ship-packages-within-d-days/1056-capacity-to-ship-packages-within-d-days.cpp
class Solution {
private:
    int findMin(vector<int> & weights, int weight){
        int minDays = 1;
        int minWeight = 0;
        for(int i=0; i<weights.size(); i++){
            if(minWeight + weights[i] <= weight){
                minWeight += weights[i];
            }
            else{
                minDays++;
                minWeight = weights[i];
            }
        }
        return minDays;
    }
public:
    int shipWithinDays(vector<int>& weights, int days) {
        int low = *max_element(weights.begin(),weights.end());
        int high = accumulate(weights.begin(),weights.end(),0);
        while(low <= high){
            int mid = low + (high - low) / 2;
            int val = findMin(weights,mid);
            if(val <= days){
                high = mid -1;
            }
            else{
                low = mid + 1;
            }
        }
        return low;
    }
};

File: 1056-capacity-to-ship-packages-within-d-days/1056-capacity-to-ship-packages-within-d-days.java
class Solution {
    private int findDays(int[] weights,int weight){
        int sum = 0, days = 1;
        for(int i=0; i<weights.length; i++){
            if(sum + weights[i] > weight){
                days++;
                sum = 0;
            }
            sum+=weights[i];
        }
        return days;
    }
    public int shipWithinDays(int[] weights, int days) {
         int low = Arrays.stream(weights).max().getAsInt();
        int high = Arrays.stream(weights).sum();
        while(low <= high){
            int mid = low + (high - low)/2;
            int requiredDay = findDays(weights,mid);

            if(requiredDay > days){
                low = mid+1;
            }
            else high = mid-1;
        }
        return low;// low is started in the notpossible side;
    }
        
 }
